lib/expression/set.go

```diff
@@ -18,65 +18,31 @@
 
 package expression
 
-import (
-	"golang.org/x/exp/maps"
-)
+import "github.com/gravitational/teleport/lib/utils"
 
-// Set is a map of type string key and struct values.
-type Set map[string]struct{}
-
-// NewSet returns Set from given string slice.
-func NewSet(values ...string) Set {
-	s := make(Set, len(values))
-	for _, value := range values {
-		s[value] = struct{}{}
-	}
-	return s
-}
-
-func (s Set) add(values ...string) Set {
-	out := s.clone()
-	for _, value := range values {
-		out[value] = struct{}{}
-	}
-	return out
-}
-
-func (s Set) contains(str string) bool {
-	_, ok := s[str]
-	return ok
+// Set is a map of type string key and struct values. Set is a thin wrapper over
+// the utils.Set[T] generic set type, allowing the Set to implement the
+// interface(s) required for use with the expression package.
+type Set struct {
+	utils.Set[string]
 }
 
-func (s Set) remove(values ...string) any {
-	out := s.clone()
-	for _, value := range values {
-		delete(out, value)
-	}
-	return out
-}
-
-func (s Set) transform(f func(string) string) Set {
-	out := make(Set, len(s))
-	for str := range s {
-		out[f(str)] = struct{}{}
-	}
-	return out
-}
-
-func (s Set) clone() Set {
-	return maps.Clone(s)
+// NewSet constructs a new set from an arbitrary collection of elements
+func NewSet(values ...string) Set {
+	return Set{utils.NewSet(values...)}
 }
 
-func (s Set) items() []string {
-	return maps.Keys(s)
+// remove creates a new Set containing all values in the receiver Set, minus
+// all supplied elements. Implements expression.Remover for Set.
+func (s Set) remove(elements ...string) any {
+	return Set{s.Set.Clone().Remove(elements...)}
 }
```

Comment:
I'm guessing this method needs to stay here so we're able to provide string arguments directly from the expressions parser?

Reply:
Yeah. And it's part of the `remover` interface used by the parser to work on `Dict`s and `Set`s interchangeably, which is only defined inside the `expression` package.